server/src/controllers/NotificationController.ts

```diff
@@ -0,0 +1,108 @@
+import { Request, Response } from 'express';
+import webpush from 'web-push';
+import dotenv from 'dotenv';
+import User from '../models/User';
+
+dotenv.config();
+
+webpush.setVapidDetails(
+  'https://cs472-roadwatch.vercel.app/', // TODO: Replace with final registered domain (e.g., https://roadwatch.com )
+  process.env.NOTIFICATION_PUBLIC_KEY as string,
+  process.env.NOTIFICATION_PRIVATE_KEY as string
+);
+
+interface ISendNotificationRequestBody {
+  id: string;
+  title: string;
+  options?: NotificationOptions;
+}
+
+export const saveSubscription = async (req: Request, res: Response) => {
+  const { id, subscription } = req.body;
+
+  if (!id) {
+    return res.status(401).send({
+      error: 'Unauthorized',
+    });
+  }
+
+  try {
+    await User.findByIdAndUpdate(id, {
+      $addToSet: { notificationSubscriptions: subscription },
+    });
+
+    res.statusMessage = 'User successfully subscribed to notifications.';
+    res.sendStatus(200);
+  } catch (error) {
+    console.error('Notification Subscription Error: ', error);
+    return res.sendStatus(400);
+  }
+};
+
+// TODO: Create `sendNotificationToCommunity` endpoint.
+
+export const unsubscribe = async (
+  req: Request<object, object, { id: string; subscription: PushSubscription }>,
+  res: Response
+) => {
+  const { id, subscription } = req.body;
+
+  if (!id) {
+    return res.status(401).send({
+      error: 'Unauthorized',
```

Comment:
This seems to be trying to do some sort of authorization check based on the id being passed in. This is not how user authorization is done and also this wouldn't work. 

Say for example I hit this endpoint with postman. All this checks is whether I provided an id or not, it could be gibberish and it would still allow me to proceed past this. If you want this to only be hit by authorized users then hit the validation middleware before entering here

Reply:
Good point. I'll modify this. I was having some difficulty using the `validateToken` function, so I'll let you know if I need any help.